### PR TITLE
Add preprocessing step for transfo-xl tokenization to avoid tokenizing words followed by punction to <unk>

### LINKED ISSUE #2000: Wrong tokenization in Transformer-XL documentation
## 🐛 Bug

This is a documentation-related bug. In the [TransfoXL documentation](https://huggingface.co/transformers/model_doc/transformerxl.html), the tokenization example is wrong. The snippet goes:
```
tokenizer = TransfoXLTokenizer.from_pretrained('transfo-xl-wt103')
...
input_ids = torch.tensor(tokenizer.encode("Hello, my dog is cute")).unsqueeze(0)  # Batch size 1
```
This code outputs the tokens `[24, 617, 3225, 23, 16072]`, of which `24` is `<unk>`.

The problem comes from the fact that Transformer-XL does **not** use a wordpiece vocabulary, but a regular (whole-word) one. Also, in WT-103, punctuation marks are split from the words. Consequently, the example should read instead (note that space in from of `,`):
```
tokenizer = TransfoXLTokenizer.from_pretrained('transfo-xl-wt103')
...
input_ids = torch.tensor(tokenizer.encode("Hello , my dog is cute")).unsqueeze(0)  # Batch size 1
```

It would also be nice to warn the user about this fact in the documentation, perhaps in `TransfoXLTokenizer`'s docstring?

### PR DESCRIPTION
**Pull request by patrickvonplaten**
The problem is well shown in Issue #2000 .
This PR adds a preprocessing step to transfo-xl tokenization to better deal with the problem.

### CODE REVIEW

File: src/transformers/tokenization_transfo_xl.py [hunk 1]
```diff
@@ -295,6 +297,19 @@ def _tokenize(self, line, add_eos=False, add_double_eos=False):
...
 
+    def prepare_for_tokenization(self, text, **kwargs):
+
+        look_ahead_for_special_token = '(?=[{}])'.format(self.punctuation_symbols)
+        look_ahead_to_match_all_except_space = '(?=[^\s])'  # noqa: W605
+        punction_without_space_before = '[^\s][{}]'.format(self.punctuation_symbols)  # noqa: W605
+        # add spaces before punctuation symbols as should be done in transfo-xl
+        if "add_space_before_punct_symbol" in kwargs and kwargs["add_space_before_punct_symbol"]:
+            text = re.sub(r'' + look_ahead_for_special_token + look_ahead_to_match_all_except_space, r' ', text)
+        elif re.search(punction_without_space_before, text):
```
Comment (patrickvonplaten) [inline src/transformers/tokenization_transfo_xl.py]: If the `elif` statement would slow the tokenizer down too much for long texts that are "correctly" preprocessed, another possibility would be to simply add a warning every time transfo-xl tokenizier is used. What do you think?
@mfuntowicz , @thomwolf

Comment (mfuntowicz) [reply src/transformers/tokenization_transfo_xl.py]: I would not put to much concern in the speediness of Python's tokenizers, they inherently slow. Also, TransfoXL already has it's fast counterpart, so it might mitigate the need for Python's one to be fast.

@patrickvonplaten: Did you observe the same issue with `TranformerXlTokenizerFast` ? If the issue is also present in the later, we need to have a look quickly on Rust side, as it'll be picked by default by AutoTokenizer.

**Review by mfuntowicz — APPROVED** [review]

File: src/transformers/tokenization_transfo_xl.py (hunk 1 above)
Comment (BramVanroy) [reply src/transformers/tokenization_transfo_xl.py]: Just a thought: you are now creating the three regexes again and again on each call to prepare_for_tokenization(). (Compiling) regex is slow (or at least that's what I remember from a long time ago when I was switched Perl code to Python). Might it be an option to compile the regex into a regex object as a class instance attribute? That way the regex pattern just needs to be compiled once (in init).

Comment (patrickvonplaten) [reply src/transformers/tokenization_transfo_xl.py]: @BramVanroy - thanks! Just changed it to compilation at instantiation time

Comment (patrickvonplaten) [reply src/transformers/tokenization_transfo_xl.py]: @mfuntowicz 
I observe the same issue when using `TranformerXlTokenizerFast`.

**Review by LysandreJik — APPROVED** [review]
This works for me. I agree with @mfuntowicz that optimizing for speed should be left to the rust tokenizer; this is understandable by users and shouldn't slow down the tokenization process too much.

**Comment by patrickvonplaten** [comment]
Good to be merged then, I think!

**Comment by summerstay** [comment]
Using this new preprocessing step, what is the proper way to format a paragraph of text so that it will work as a prompt for XLNet? Just a few lines of code as an example would make it clear to me, I think. 

The "padding text" in the example contains sentences with punctuation not separated by spaces. At the end of the text is "<eod> </s> <eos>" Shouldn't I include a <eos> after every sentence, by hand? If not and it can be done automatically, how do you avoid things like Mr. or Dr. that end in periods but don't end the sentence? 

Thanks.

**Comment by patrickvonplaten** [comment]
Hi @summerstay, this preprocessing step is only necessary for Transfo-XL Net. For XLNet, you don't need to separate the text. Try out this code to see what I mean:

```
def show_tokenization(text, model_name):
    tok = AutoTokenizer.from_pretrained(model_name)
    print(tok.decode(tok.encode(text, add_special_tokens=False)))


show_tokenization('This is an example. See what happens with, and. ?', 'transfo-xl-wt103')
show_tokenization('This is an example. See what happens with, and. ?', 'xlnet-base-cased')

# prints:
# You might want to consider setting `add_space_before_punct_symbol=True` as an argument to the `tokenizer.encode()` to avoid tokenizing words with punctuation symbols to the `<unk>` token
# This is an <unk> See what happens <unk> <unk>?
# This is an example. See what happens with, and.?
```
